doc/developer/design/20230330_recursion_limit.md

```diff
@@ -0,0 +1,159 @@
+
+Alternatively, we could set the limits in a system / session variable. Note that neither of these would be suitable for the 3. use case.
+
+A system variable applies to all queries in the entire system. (Even if we make a dataflow pick up the current value only at dataflow creation, this can affect all queries if a restart suddenly happens.) Therefore, a system variable is not suitable for the 2. and 3. use cases. However, it is suitable for 1. (the hard limit).
+
+[Most system variables can't be modified by a user on her own, but only by support.](https://materialize.com/docs/sql/show/#system-variables) However, [there is a precedent](https://materializeinc.slack.com/archives/C02FWJ94HME/p1680209826137869?thread_ts=1680209764.706779&cid=C02FWJ94HME) for a system variable that is modifiable by users. As mentioned above, we would like the hard limit to be easily modifiable by users, and therefore if we decide to put the hard limit in a system variable, we should definitely make it modifiable by users.
+
+Session variables have the problem that a custom value will be lost across environmentd restarts, meaning that it is impossible to install a view where the limit is different from the default, and have the custom limit stick between restarts. (We could conceivably add the value to the catalog along with the query's SQL definition, but this is probably not worth the trouble.) Therefore, a session variable wouldn't be suitable for the 1. (the hard limit) and 3. use cases, because for these one has to raise the limit for a production view in a manner that survives restarts. It might be kind of ok for the 2. use case (debugging): one can temporarily change the limit, and run the query that is being debugged. However, there is still the risk that somebody forgets to change the limit back. Also note that this wouldn't allow for different limits in different WMR blocks inside the same query.
+
+In general, it's inadvisable for system / session variables to affect query semantics in a major way, so we rejected this option.
+
+### Query-level option
+
+As discussed [above](#syntax), we would like to make it possible to set different limits for different WMR blocks inside the same query, which wouldn't be possible with a query-level option.
+
+Also note that we still wouldn't avoid adding a new options clause, because the current SELECT-level options clause is not query-level.
+
+### WHERE clause with a special, unmaterializable function
+
+Theoretically, we could put the limit in WHERE clauses with a special, unmaterializable function, e.g., `WHERE iteration_number() < 100`. However, this would have the difficulty that `PredicatePushdown` would happily push `iteration_number() < 100` all the way to the sources, and therefore out of the recursion. We could of course change `PredicatePushdown` to have a special case for expressions involving `iteration_number()`, but the problem is that it's hard to know what other transforms might have similar problems.
```

Comment:
Our current `UnmaterializableFunc` enum uses the term "unmaterializable" as a substitute for "depends on the read-only `environmentd` context (that is, they can be substituted at query optimization / compilation time). However the proposed `iteration_number()` function goes even further, as it depends on the execution context of the rendered dataflow (so the value can only be substituted at runtime). This entails that our functions `eval` logic will need to be augmented with some notion of a runtime context. Maybe worth noting this here as a known complication.